04-onion-routing.md

```diff
@@ -518,20 +552,17 @@ At this point, the processing node can generate a _rho_-key and a _gamma_-key.
 
 The routing information is then deobfuscated, and the information about the
 next hop is extracted.
-To do so, the processing node copies the `hops_data` field, appends 65 `0x00`-bytes,
-generates 1365 pseudo-random bytes (using the _rho_-key), and applies the result
-,using `XOR`, to the copy of the `hops_data`.
-The first 65 bytes of the resulting routing information become the `per_hop`
-field used for the next hop. The next 1300 bytes are the `hops_data` for the
-outgoing packet.
+To do so, the processing node copies the `hop_payloads` field, appends 1300 `0x00`-bytes,
+generates `2*1300` pseudo-random bytes (using the _rho_-key), and applies the result, using `XOR`, to the copy of the `hop_payloads`.
+The first few bytes correspond to the varint-encoded length `l` of the `hop_payload`, followed by `l` bytes of the resulting routing information become the `hop_payload`, and the 32 byte HMAC.
+The next 1300 bytes are the `hop_payloads` for the outgoing packet.
 
-A special `per_hop` `HMAC` value of 32 `0x00`-bytes indicates that the currently
-processing hop is the intended recipient and that the packet should not be forwarded.
+A special `HMAC` value of 32 `0x00`-bytes indicates that the currently processing hop is the intended recipient and that the packet should not be forwarded.
```

Comment:
That would create yet another special case. We can maintain a stronger separation of layers b signalling termination at the onion layer, rather than at the "application" layer, otherwise TLV code is _required_ in order to test if one's Sphinx construction/processing code is compliant.